presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -1171,6 +1171,33 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
     {
         ImmutableList.Builder<PartitionUpdate> partitionUpdatesForMissingBucketsBuilder = ImmutableList.builder();
         HiveStorageFormat storageFormat = table.getPartitionColumns().isEmpty() ? handle.getTableStorageFormat() : handle.getPartitionStorageFormat();
+
+        // empty unpartitioned bucketed table
+        if (table.getPartitionColumns().isEmpty() && partitionUpdates.isEmpty()) {
```

Comment:
Is it possible to emit a `PartitionUpdate` in `HivePageSink#doFinish` when no data is written for unpartitioned table, similar to https://github.com/prestodb/presto/blob/e629ad1879316b010df3c31db6ec259dc55c35b9/presto-hive/src/main/java/com/facebook/presto/hive/CreateEmptyPartitionProcedure.java#L109-L119 ?

You can check whether it's a partitioned table by `partitionColumnsInputIndex.length == 0`

Reply:
> Is it possible to emit a PartitionUpdate in HivePageSink#doFinish when no data is written for unpartitioned

It only makes sense when writing a bucketed table. Empty files special case is already handled here. So I'm not sure if this is a good idea of splitting this responsibility into 2 different places.

Also sending an empty partition update for every writer seems to be inefficient, as it requires every writer to send one.

Reply:
Let's talk in person. 

I view it as two separate responsibilities:
1. Report partitions written into (unpartitioned table is modeled as a special partition)
2. Create empty files for bucketed partition. 

So , I view it more as, we didn't report the partition in `HivePageSink`, which is sink's responsibility.

Reply:
> So , I view it more as, we didn't report the partition in HivePageSink, which is sink's responsibility.

Technically the current behaviour is correct. If the HivePageSink did't write a partition - it won't be reported. So if the `HivePageSink` didn't write to the `UNPARTITIONED` partitions, it shouldn't report it.

Reply:
Per offline discussion, there are two ways to create a bucketed unpartitioned table:

1. CREATE TABLE
2. CREATE TABLE AS

Emit empty `PartitionUpdate` only solve the second case but not the first. As a result, we still need to handle it in `finishCreate`/`finishInsert`